Add Mazda Connected Services integration

bdr99 opened the pull request:
<!--
  You are amazing! Thanks for contributing to our project!
  Please, DO NOT DELETE ANY TEXT from this template! (unless instructed).
-->

## Proposed change
<!--
  Describe the big picture of your changes here to communicate to the
  maintainers why we should accept this pull request. If it fixes a bug
  or resolves a feature request, be sure to link to that issue in the
  additional information section.
-->
This PR adds a new integration for Mazda Connected Services, a connected vehicle platform which allows for remote monitoring and control of Mazda vehicles through a mobile app for iOS and Android.

Currently, I've only implemented the Sensor platform, but other platforms like Lock and Device Tracker can be added in the future, as well as services to start the engine and flash the lights.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [ ] Bugfix (non-breaking change which fixes an issue)
- [x] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #
- This PR is related to issue: 
- Link to documentation pull request: https://github.com/home-assistant/home-assistant.io/pull/16365

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [x] The code change is tested and works locally.
- [x] Local tests pass. **Your PR cannot be merged unless tests pass**
- [x] There is no commented out code in this PR.
- [x] I have followed the [development checklist][dev-checklist]
- [x] The code has been formatted using Black (`black --fast homeassistant tests`)
- [x] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [x] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [x] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [x] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [x] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

bdr99 committed:
Add Mazda Connected Services integration

bdr99 committed:
Increase test coverage

frenck reviewed (changes requested):
This looks really good!

Some small comments to look into. Also, a couple of lines aren't covered by tests. Especially the ones in the config flow that need to be covered (as we require 100% coverage in the configuration flow).

Good job 👍

bdr99 commented on homeassistant/components/mazda/config_flow.py:
```diff
@@ -0,0 +1,123 @@
+                await self.async_set_unique_id(user_input[CONF_EMAIL])
+
+                for entry in self._async_current_entries():
+                    if entry.unique_id == self.unique_id:
```
That's a good point, but I'm not sure what to do about it. How can I find the current entry and update it?

frenck replied:
Right... I'm ok with it personally. Nevertheless, I'm going to leave this op for a second opinion from a fellow reviewer if you don't mind 😄

balloob replied:
Nothing we can do. Also reauth would never pass :)

bdr99 commented on homeassistant/components/mazda/strings.json:
```diff
@@ -0,0 +1,35 @@
+            }
+        }
+    },
+    "title": "Mazda Connected Services"
```
Looking at the Android app source code, it looks like the name is different in other countries. Therefore I believe it should be translatable.

bdr99 committed:
Code review fixes

bdr99 commented on homeassistant/components/mazda/__init__.py:
```diff
@@ -0,0 +1,167 @@
+    ) as ex:
+        _LOGGER.error("Error occurred during Mazda login request: %s", ex)
+        raise ConfigEntryNotReady from ex
+    except Exception as ex:
```
OK, I removed this catch.

bdr99 commented on homeassistant/components/mazda/config_flow.py:
```diff
@@ -0,0 +1,123 @@
+        errors = {}
+
+        if user_input is not None:
+            await self.async_set_unique_id(user_input[CONF_EMAIL])
```
Nice catch. I changed it so it converts to lower case.

bdr99 commented:
@frenck Thanks for the review! I just pushed a commit that addresses most of your comments. Not sure what to do about the email address changing though. Do you have any ideas?

frenck reviewed (approved):
I think this looks great! 👍 

Since this is a new integration, I'm going to ask a second reviewer for a quick look as well.

balloob commented on homeassistant/components/mazda/__init__.py:
```diff
@@ -0,0 +1,161 @@
+        try:
+            async with async_timeout.timeout(10):
+                vehicles = await mazda_client.get_vehicles()
+                for vehicle in vehicles:
+                    status = await mazda_client.get_vehicle_status(vehicle["id"])
```
You're doing an await for each vehicle but it's inside a timeout block for 10 seconds. The odds of someone hitting this gets higher if they have more cars. We should do a timeout per request. Also, not sure about API rate limits, but if possible, try to get all vehicle statuses at once. 

If you don't want to ddos the API if someone has 100 cars but still provide maximum parallelism, you can use `gather_with_concurrency` from `homeassistant.util.async_`.

Something like this:

```python
# Not tested, from memory :)

async def with_timeout(task):
    async with async_timeout.timeout(10):
        return await task

async with async_timeout.timeout(10):
    vehicles = await mazda_client.get_vehicles()

statuses = await gather_with_concurrency(5, [
    with_timeout(mazda_client.get_vehicle_status(vehicle["id"]))
    for vehicle in vehicles
])

for vehicle, status in zip(vehicles, statuses):
    vehicle['status'] = status
```

bdr99 replied:
That's a great point, I didn't consider this. I updated the code to use `gather_with_concurrency`.

balloob commented on homeassistant/components/mazda/config_flow.py:
```diff
@@ -0,0 +1,123 @@
+
+                for entry in self._async_current_entries():
+                    if entry.unique_id == self.unique_id:
+                        self.hass.config_entries.async_update_entry(
+                            entry, data=user_input
+                        )
+
+                        # Reload the config entry otherwise devices will remain unavailable
+                        self.hass.async_create_task(
+                            self.hass.config_entries.async_reload(entry.entry_id)
+                        )
```
You've re-implemented the exact logic of [`self._abort_if_unique_id_configured`](https://github.com/home-assistant/core/blob/dev/homeassistant/config_entries.py#L898) so use that method instead.

bdr99 replied:
It looks like `self._abort_if_unique_id_configured` aborts with reason `"already_configured"`. But I am doing reauth, so I need to abort with reason `"reauth_successful"` in order to display the correct message to the user, right? Unless I'm missing something here.

balloob replied:
You're right. My bad.

balloob commented on homeassistant/components/mazda/config_flow.py:
```diff
@@ -0,0 +1,123 @@
+
+        if (
+            user_input is not None
+            and user_input.get(CONF_EMAIL) is not None
+            and user_input.get(CONF_PASSWORD) is not None
+            and user_input.get(CONF_REGION) is not None
```
The form schema specifies these values as required and so they will always be there.

balloob commented on homeassistant/components/mazda/manifest.json:
```diff
@@ -0,0 +1,9 @@
+  "documentation": "https://www.home-assistant.io/integrations/mazda",
+  "requirements": ["pymazda==0.0.8"],
+  "codeowners": ["@bdr99"],
+  "quality_scale": "platinum"
```
Niiice 🙇

balloob commented on homeassistant/components/mazda/sensor.py:
```diff
@@ -0,0 +1,264 @@
+
+    entities = []
+
+    if coordinator.data is not None:
```
This can't happen because you raise a `ConfigEntryNotReady` when the first update request fails.

balloob commented on homeassistant/components/mazda/sensor.py:
```diff
@@ -0,0 +1,264 @@
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        vehicle_name = self.get_vehicle_name()
```
Can the vehicle name change dynamically? Or is this always the same? 

If it's always the same, you'll get slightly better performance by calculating the name and the unique ID in the constructor. You could take `name_part` and `unique_id` part in the base class and have each entity pass their unique part in the constructor.

bdr99 replied:
Actually, the vehicle name can change dynamically. The user can rename their vehicle in the MyMazda app at any time, which would cause the name returned by the API to be different.

balloob reviewed (approved):
Looks awesome! Two tiny comments. 👍 amazing one 🙇

balloob commented on homeassistant/components/mazda/config_flow.py:
```diff
@@ -0,0 +1,123 @@
+
+        if user_input is not None:
+            await self.async_set_unique_id(user_input[CONF_EMAIL].lower())
+            self._abort_if_unique_id_configured()
```
If the user initiates a config flow, we should not abort if a unique ID is configured. We do this so we can also override ignored entries.

Home Assistant will automatically remove the old one and have the new one replace it.

bdr99 committed:
More code review fixes

balloob commented on homeassistant/components/mazda/__init__.py:
```diff
@@ -0,0 +1,173 @@
+
+            return vehicles
+        except MazdaAuthenticationException as ex:
+            hass.async_create_task(
```
Thinking now, we might want to just do a reload of the config entry here. Once we merge #45528, all entities will be marked automatically as `unavailable` while your `async_setup_entry` logic will trigger a re-auth and no longer try to make a request every 60 seconds.

balloob commented:
So going to merge this. It's great! My thought above can be addressed in a new PR. Probably better to see the effect of it once #45528 is actually in. 

Great first contribution 💯 🥇

MartinHjelmare commented on tests/components/mazda/test_init.py:
```diff
@@ -0,0 +1,100 @@
+    assert len(entries) == 1
+    assert entries[0].state == ENTRY_STATE_LOADED
+
+    coordinator = hass.data[DOMAIN][config_entry.entry_id][DATA_COORDINATOR]
```
Please don't access integration details in the tests. Following that practice will make the tests more robust. Here instead move time forward to force a coordinator update. We have this helper:

https://github.com/home-assistant/core/blob/61a987061e9da549ccdf466a91869af0c3ab2a5f/tests/common.py#L285-L286

MartinHjelmare commented on tests/components/mazda/test_init.py:
```diff
@@ -0,0 +1,100 @@
+
+    await hass.config_entries.async_unload(entry.entry_id)
+    await hass.async_block_till_done()
+    assert not hass.data.get(DOMAIN)
```
Assert the entry state instead. It should be `ENTRY_STATE_NOT_LOADED`.

MartinHjelmare reviewed (commented):
Very nice! Some small comments to address in a future PR. :+1:
